src/resolvers/libsodium.rs

```diff
@@ -0,0 +1,312 @@
+/// via [sodiumoxide](https://crates.io/crates/sodiumoxide).
+#[derive(Default)]
+pub struct SodiumResolver;
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<dyn Random>> {
+        Some(Box::new(SodiumRng::default()))
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<dyn Dh>> {
+        match *choice {
+            DHChoice::Curve25519 => Some(Box::new(SodiumDh25519::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<dyn Hash>> {
+        match *choice {
+            HashChoice::SHA256 => Some(Box::new(SodiumSha256::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<dyn Cipher>> {
+        match *choice {
+            CipherChoice::ChaChaPoly => Some(Box::new(SodiumChaChaPoly::default())),
+            _ => None,
+        }
+    }
+}
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
+}
+
+impl SodiumDh25519 {
+    fn convert_to_private_key(key: &mut [u8; 32]) {
+        key[0] &= 248;
+        key[31] &= 127;
+        key[31] |= 64;
+    }
+}
+
+impl Default for SodiumDh25519 {
+    fn default() -> SodiumDh25519 {
+        sodiumoxide::init().unwrap();
+
+        SodiumDh25519 {
+            privkey: sodium_curve25519::Scalar([0; 32]),
+            pubkey: sodium_curve25519::GroupElement([0; 32]),
+        }
+    }
+}
+
+impl Dh for SodiumDh25519 {
+    fn name(&self) -> &'static str {
+        "25519"
+    }
+
+    fn pub_len(&self) -> usize {
+        32
+    }
+
+    fn priv_len(&self) -> usize {
+        32
+    }
+
+    fn set(&mut self, privkey: &[u8]) {
+        self.privkey = sodium_curve25519::Scalar::from_slice(privkey)
+            .expect("Can't construct private key for Dh25519");
+        self.pubkey = sodium_curve25519::scalarmult_base(&self.privkey);
+    }
+
+    fn generate(&mut self, rng: &mut dyn Random) {
+        let mut privkey_bytes = [0; 32];
+        rng.fill_bytes(&mut privkey_bytes);
+
+        Self::convert_to_private_key(&mut privkey_bytes);
+
+        self.privkey = sodium_curve25519::Scalar::from_slice(&privkey_bytes)
+            .expect("Can't construct private key for Dh25519");
+        self.pubkey = sodium_curve25519::scalarmult_base(&self.privkey);
+    }
+
+    fn pubkey(&self) -> &[u8] {
+        &self.pubkey[0..32]
+    }
+
+    fn privkey(&self) -> &[u8] {
+        &self.privkey[0..32]
+    }
+
+    fn dh(&self, pubkey: &[u8], out: &mut [u8]) -> Result<(), ()> {
+        let pubkey = sodium_curve25519::GroupElement::from_slice(&pubkey[0..32])
+            .expect("Can't construct public key for Dh25519");
+        let result = sodium_curve25519::scalarmult(&self.privkey, &pubkey);
+
+        match result {
+            Ok(ref buf) => {
+                out[..32].copy_from_slice(&buf[0..32]);
```

Comment:
I know it's non-canonical Rust, but let's use the `copy_slices!` macro for consistency until we change it library-wide in one fell swoop:

```rust
copy_slices!(&buf, out);
```

There are a couple more instances of this below.